From 3016da424e7b6c9bc2a27e0e15bab6644dfbfe69 Mon Sep 17 00:00:00 2001
From: "Dmitry.Neverov" <dmitry.neverov@jetbrains.com>
Date: Mon, 22 Aug 2022 11:24:23 +0200
Subject: [PATCH] [dbg] handle connection reset from IDE

So that when client is killed, cmake terminates as well instead of trying to
send errors to the disconnected socket indefinitely.

diff --git a/Source/Sysprogs/BasicIncomingSocket.h b/Source/Sysprogs/BasicIncomingSocket.h
index 12edc14324..090a4606f9 100644
--- a/Source/Sysprogs/BasicIncomingSocket.h
+++ b/Source/Sysprogs/BasicIncomingSocket.h
@@ -3,11 +3,19 @@
 #ifdef _WIN32
 #include <WinSock2.h>
 typedef int socklen_t;
+
+bool isDisconnected() {
+	return WSAGetLastError() == WSAECONNRESET;
+}
 #else
 #include <sys/socket.h>
 #include <netinet/ip.h>
 typedef int SOCKET;
 static void closesocket(SOCKET socket) { close(socket); }
+
+bool isDisconnected() {
+	return errno == ECONNRESET;
+}
 #endif
 
 /*
@@ -20,8 +28,12 @@ private:
 	SOCKET m_Socket, m_AcceptedSocket;
 
 public:
+	bool m_Disconnected;
+
 	BasicIncomingSocket(int tcpPort) : m_Socket(0), m_AcceptedSocket(0)
 	{
+		m_Disconnected = false;
+
 		sockaddr_in addr = {
 			AF_INET,
 		};
@@ -86,6 +98,13 @@ public:
 		if (m_AcceptedSocket <= 0)
 			return false;
 		int done = send(m_AcceptedSocket, (const char *)pData, size, 0);
+		if (done == -1) {
+			if (isDisconnected()) {
+				m_Disconnected = true;
+				cmSystemTools::Error("Socket disconnected.");
+				cmSystemTools::SetFatalErrorOccurred();
+			}
+		}
 		return done == size;
 	}
 
diff --git a/Source/Sysprogs/HLDPServer.cpp b/Source/Sysprogs/HLDPServer.cpp
index 3f8a0f813d..c74dae59fd 100644
--- a/Source/Sysprogs/HLDPServer.cpp
+++ b/Source/Sysprogs/HLDPServer.cpp
@@ -841,6 +841,10 @@ namespace Sysprogs
 
 		for (;;)
 		{
+			if (m_pSocket->m_Disconnected)
+			{
+				break;
+			}
 			builder.Reset();
 
 			RequestReader reader;
-- 
2.50.1 (Apple Git-155)

